### ISSUE TITLE
Pool issues

### ISSUE THREAD

**Issue by ysbaddaden** [crystal-lang/crystal-db#77]
With default values for the pool:
1. DB can create unlimited connections to the database;
2. DB may only keep 1 idle connection;
3. DB will disconnect connections on release (and keep 1 connection alive).

Since the default is to checkout/query/release, we easily end up with *many* connect / disconnect whenever we stress an application a little bit.

Maybe I'm wrong, but releasing a connection only checks `max_idle_pool_size` and doesn't seem to verify whether there are pending checkouts, which means that the connection wouldn't become idle.

In fact, I don't understand why the connection should be closed immediately. Why not push it back into the pool, and have a reaper that will close connections to fall below `max_idle_pool_size` after an idle threshold of some sort? That would avoid countless (dis)connect while still reducing the number of actual idle connections when the system is no longer under stress.

**Comment by ysbaddaden** [comment]
After some debugging I found out that `@total.size` grows **far larger** than `max_pool_size`, and is only limited to how many connections are available on the SQL server.

For example with `max_pool_size=15&max_pool_idle_size=15&retry_attemps=5`; I have 300 incoming parallel requests that issue at most 5 SQL requests within a `db.using_connection` block. Debugging `@total.size` within `can_increase_pool_size`, it will stay at 1 or 2 for some time, then suddenly jump to 152, 175 then I start having `DB::ConnectionRefused` exceptions because my SQL server won't accept more than 200 incoming connections.

**Comment by ysbaddaden** [comment]
There is a race condition in `DB#checkout` and `DB#build_resource` when called many times concurrently from many fiber.

`#build_ressource` calls `@factory.call` which create & connect a TCPSocket and possibly handshake... which will trigger the event loop, but we only push to `@total` *after* the factory returns, leaving an open window for a race condition in `#checkout`.

There must either be a lock in `#checkout` or keeping `@total_size` and `@available_size` as atomic values that will be incremented/decremented, but incapable to increment larger than maximum values, thus only being capable to increment `@total.size` would call `#build_resource` safely.

**Comment by ysbaddaden** [comment]
In the meantime, the issue is avoided with `initial_pool_size  = max_pool_size = max_idle_pool_size`.

**Comment by RX14** [comment]
Obviously the first step is benchmarking this outside framework benchmarks without any http code.

**Comment by ysbaddaden** [comment]
@jwoertink yes, that was my workaround.

**Comment by jwoertink** [comment]
Ok, that change didn't fix my error. Actually, it sort of got worse on my end. We let it sit for a day, and it seems all the connections just start dying in the pool.

**Comment by Xosmond** [comment]
@ysbaddaden Is it ok if when I set max connections to 25, I got sometimes 90 or 300 connections to the db when stressing? And I got the error on the log "Exception: Too many connections"

**Comment by Xosmond** [comment]
@ysbaddaden Is there any way I can help? Or some other language implementation could help help?

**Comment by Xosmond** [comment]
Sometimes the database close idle connections, so when crystal-db tries to use one of them, we got a `Broken pipe error` We should have a way to catch this and create a new connection. But I think we should fix the problem that create more connections that the limit first.

**Comment by ysbaddaden** [comment]
Connections dying is another topic than this issue (race condition that can create lots of connections suddenly). Ideally the driver (or the pool) should catch that a connection has been closed (EPIPE, ECONNRESET, ...) should clean the connection, (try to) reconnect then run the query again, transparently.

**Comment by Xosmond** [comment]
@asterite I tried removing the build_resource from checkout method but it looks like that is where all the connections are made. Because once removed, no connection could be made to db. So looks like the initializer method is not actually creating the connections. I'm I wrong?

**Comment by bcardiff** [comment]
@Xosmond the initial connections are created at:

https://github.com/crystal-lang/crystal-db/blob/af916fbffb739c0d0231adb61756e5700cb454e0/src/db/pool.cr#L17

And the `build_resource` uses the default factory which is defined at

https://github.com/crystal-lang/crystal-db/blob/af916fbffb739c0d0231adb61756e5700cb454e0/src/db/database.cr#L50-L54

**Comment by Xosmond** [comment]
@bcardiff Changing this block:

https://github.com/crystal-lang/crystal-db/blob/af916fbffb739c0d0231adb61756e5700cb454e0/src/db/pool.cr#L31-L46

to this:
```crystal
def checkout : T
    resource = if @available.empty?
      wait_for_available
      pick_available
    else
      pick_available
    end

  @available.delete resource
  resource.before_checkout
  resource
end
```
This change made that no connections to db could be made, so I think that is why setting `initial_pool_size=5&max_pool_size=5&max_idle_pool_size=5` doesn't work as @jwoertink said. I also tested it and on high demand it doesn't work. So I think all connections are being done in that line. I see the code but the testing says another thing

**Comment by bcardiff** [comment]
Without some description of how the pool is been used is hard to follow. I'm also not sure if you are still referring to the connections dying issue (that should be discussed in another issue) or to the race condition this issue refers.

Whatever effort is done, I would like to have some script for testing, at least manually so experimentation can be shared fully.

**Comment by Xosmond** [comment]
@bcardiff Sure I'll let you know

**Comment by bcardiff** [comment]
Hey, I've just submitted a PR that should fix this issue. Feedback is welcome!
